lib/src/piece/piece.dart

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import '../back_end/code_writer.dart';
+
+/// Base class for the formatter's internal representation used for line
+/// splitting.
+///
+/// We visit the source AST and convert it to a tree of [Piece]s. This tree
+/// roughly follows the AST but includes comments and is optimized for
+/// formatting and line splitting. The final output is then determined by
+/// deciding which pieces split and how.
+abstract class Piece {
```

Comment:
How about `abstract interface class` since there is no inherited behavior?

Reply:
Within a single codebase, I tend to default to `extends`. That way if I do end up defining some inherited behavior in the base class (which is fairly often for me), it just works.

Reply:
I think it's an OK choice to make, but I do think this usage of `extends` prioritizes writing over reading. When I read `implements` I need look no further to understand how the class behaves - when I read `extends` I have to make sure I understand parent class behavior. I also need to keep in mind that some of the behavior is intended for consumption by other callers, and some behavior may be intended for consumption by the parent class. In this case it's a quick hop to see no behavior exists, but the `implements` would have saved me even that hop.